src/audio/kpb.c

```diff
@@ -961,17 +962,31 @@ static void kpb_init_draining(struct comp_dev *dev, struct kpb_client *cli)
 
 		} while (buff != first_buff);
 
-		/* Calculate time in clock ticks each draining event shall
-		 * take place. This time will be used to synchronize us with
-		 * an end application interrupts.
+		/* Check if we need to control draining speed.
+		 * This will be needed if KPB's history buffer is bigger
+		 * than host buffer.
 		 */
-		drain_interval = (host_period_size / bytes_per_ms) *
-				 ticks_per_ms;
-		/* In draining intervals we will fill only two periods
-		 * and give host time to read it.
-		 * This way we are safe to not overflow host buffer.
-		 */
-		period_bytes_limit = host_period_size * 2;
+		if (kpb->kpb_buffer_size > host_buffer_size) {
+			/* Calculate time in clock ticks each draining event
+			 * shall take place. This time will be used to
+			 * synchronize us with an end application interrupts.
+			 */
+			drain_interval = (host_period_size / bytes_per_ms) *
+					 ticks_per_ms;
+			/* In draining intervals we copy only two periods
+			 * and give host time to read the buffer.
+			 * This way we are safe to not overflow it.
+			 * NOTE! For very small periods XRUN may still
+			 * occur as time needed to serve interrupt becomes
+			 * an important factor.
+			 */
+			period_bytes_limit = host_period_size * 2;
+			sync_mode_on = true;
+		} else {
+			drain_interval = 0;
+			period_bytes_limit = 0;
+			sync_mode_on = false;
+		}
```

Comment:
Can we trace the draining config/speed to help with debug as I guess this can easily change depending on topology etc.

Reply:
@lgirdwood, sure, another trace log added.